draft-ietf-tls-external-psk-importer.md

```diff
@@ -219,3 +264,18 @@ The authors thank Eric Rescorla and Martin Thomson for discussions that led to t
 as well as Christian Huitema for input regarding privacy considerations of external PSKs. John Mattsson
 provided input regarding PSK importer deployment considerations.
 
+# Thwarting Selfie
+
+The Selfie attack [Selfie] relies on a misuse of the PSK interface.  The PSK interface makes the implicit assumption
+that each PSK is known only to one client and one server.  If this assumption is violated, as in the Selfie case where
+the PSK is known to two clients and two servers, then the implicit authentication of the PSK is broken.  If two distinct
+servers have the same key then they are indistinguishable to the client, and vice versa.  To avoid this difficulty you
+can use imported PSKs with extra context that distiguishes the various actors.  As long as all distinct clients and all
+distinct servers use unique contexts the Imported PSKs they produce will be cryptographically independent.
+
+For example in the Selfie scenario if all actors include their MAC address in the `ImportedIdentity.context` field the
```

Comment:
So, I'm not sure this works. Let's assume there's two peers -- Alice and Bob, with MAC addresses MA and MB, respectively -- and they have an external PSK k. The text below suggests that Alice will import k1 = KDF(k || MA) (abusing notation for clarity!), and Bob will import k2 = KDF(k || MB). Then assume Alice sends a CH with binder computed using k1 to Bob. Alice will happily accept this CH if reflected back to herself by an attacker, since it only has k1 to verify. In contrast, Bob will fail to verify the binder, since he only has k2! 

I think this needs to do more, and further split k1 into "read" and "write" roles, and then only use them as such. For example, k1_read = KDF(k || MA || "read") and k1_write = KDF(k || MA || "write"). Bob does something similar. If Alice only computes binders with the "write" key and verifies binder using her "read" key, the reflection above won't work.

All that said, this is a great amount of complexity that needs to happen inside TLS. What we need is for the PSK to never be used by more than one entity for different roles. Can we simply outlaw that with a nasty MUST? ("Nodes MUST NOT use the same imported PSKs for different roles".) @davidben, what do you think?

Reply:
I'm not sure the "read" or "write" roles quite make sense. The PSK sender and verifier need to use the same key, so the server needs to flip. That means one of the two keys just won't be used.

I also worry focusing on the MAC address will lead folks to thinking MAC addresses are somewhat special. Presumably you want the names of the node's roles in the actual application protocol. If your protocol identifies by MAC addresses, cool. If not, you want to use something else.

I took a stab at revised text. How does this look:

```
# Addressing Selfie

The Selfie attack {{Selfie}} relies on a misuse of the PSK interface.
The PSK interface makes the implicit assumption that each PSK
is known only to one client and one server. If multiple clients or
multiple servers with distinct roles share a PSK, TLS only
authenticates the entire group. A node successfully authenticates
its peer as being in the group whether the peer is another node or itself.

Applications which require authenticating finer-grained roles while still
configuring a single shared PSK across all nodes can resolve this
mismatch either by exchanging roles over the TLS connection after
the handshake or by incorporating the roles of both the client and server
into the imported PSK context string. For instance, if an application
identifies each node by MAC address, it could use the following context string.

~~~
  struct {
    opaque client_mac<0..2^16-1>;
    opaque server_mac<0..2^16-1>;
  } Context;
~~~

If an attacker then redirects a ClientHello intended for one node to a different
node, the receiver will compute a different context string and the handshake
will not complete.

Note that, in this scenario, there is still a single shared PSK across all nodes,
so each node must be trusted not to impersonate another node's role.
```

Reply:
I think my original suggestion was made clearly enough. The client would use the client MAC and the server would _also_ use the client MAC. However I think including both MAC addresses is probably better. I won't have my laptop 'til Sunday but @davidben's change looks good to me.
Maybe say something about privacy in the non DH case too?